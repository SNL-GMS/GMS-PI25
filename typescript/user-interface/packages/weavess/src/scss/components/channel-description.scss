.channel-description {
  pointer-events: all;
  padding: 2px 0.25em 0 0.25em;
  margin-left: auto;
  margin-top: auto;
  border-radius: 2px;
  &__container {
    display: flex;
    text-align: right;
    height: 100%;
    width: 100%;
    bottom: 0px;
    margin-left: auto;
    font-size: 0.7rem;
    white-space: nowrap;
  }
  &__message-container {
    display: flex;
    align-items: center;
    gap: 0.25rem;
  }
  &__message {
    line-height: 14px;
  }
  &--error {
    color: var(--weavess-error-text, white);
    background-color: var(--weavess-error-background, darkred);
    &:hover {
      filter: brightness(1.2);
    }
  }
}
