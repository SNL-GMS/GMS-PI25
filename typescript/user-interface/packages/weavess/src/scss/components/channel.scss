$waveform-split-expand-border-color-fallback: #2c74d8;

.channel {
  width: 100%;
  border-bottom: 1px solid;
  display: flex;
  contain: strict; // Performance boost. Depends on height being set inline.

  &--maskTarget::after {
    content: '';
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    position: absolute;
    border: 1px solid white;
    z-index: 100;
    backdrop-filter: brightness(120%);
    pointer-events: none;
  }

  .channel-label-container {
    position: sticky;
    left: 0;
    z-index: 4;

    &--manually-scaled .y-axis {
      --weavess-text: var(--weavess-manually-scaled, #ff00db);
    }
  }
}

.station-split-expanded .split-channel:first-child {
  border-top: 2px solid
    var(--gms-waveform-split-expand-border-color, $waveform-split-expand-border-color-fallback);
}

.split-channel {
  width: calc(100% - 4px);
  border-left: 2px solid
    var(--gms-waveform-split-expand-border-color, $waveform-split-expand-border-color-fallback);
  border-right: 2px solid
    var(--gms-waveform-split-expand-border-color, $waveform-split-expand-border-color-fallback);
}

/*
  :last-of-type selector - is applied in the un-expanded case (no elements following the last .split-channel)
  :has(+ :not(.split-channel)) - is applied in the expanded case - this finds the .split-channel element which is not followed by another .split-channel (the last one)
*/
.channel.split-channel:last-of-type,
.channel.split-channel:has(+ :not(.split-channel)) {
  border-bottom: 2px solid
    var(--gms-waveform-split-expand-border-color, $waveform-split-expand-border-color-fallback);
}

/**
 * Adds subtle highlighting to channels when hovering over their labels
 **/
.channel:has(.label:hover) {
  background: rgba(150, 150, 150, 0.05);
}
