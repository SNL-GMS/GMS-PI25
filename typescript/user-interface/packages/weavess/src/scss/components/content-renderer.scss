.contentrenderer {
  display: flex;
  height: inherit;
  flex: 1 1 auto;
  &:focus {
    outline: none;
  }
  border-right: 1px solid;

  .contentrenderer-content {
    z-index: 2;
    position: absolute;
    height: inherit;
    width: calc(100% - 184px);
    overflow: hidden;

    &--sticky {
      z-index: 5;
      pointer-events: none;
      position: sticky;
      width: calc(100vw - 184px);
      left: 184px;
      right: 0;
      overflow: unset;
    }

    &-drag-indicator {
      position: absolute;
      top: 0px;
      bottom: 0px;
      border-left: 1px solid red;
      display: none;
    }

    &-markers {
      pointer-events: none;
      position: absolute;
      width: 100%;
      height: 100%;
      top: 0;
    }
  }
}
