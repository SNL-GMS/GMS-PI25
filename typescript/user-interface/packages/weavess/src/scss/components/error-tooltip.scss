.error-tooltip {
  display: flex;
  flex-direction: column;
  gap: 0.25rem;
  &__title {
    font-weight: bold;
    color: darkred;
  }
}
