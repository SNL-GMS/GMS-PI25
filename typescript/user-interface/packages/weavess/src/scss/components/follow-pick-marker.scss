/* Offset used to space top or bottom label and vertical marker */
$labelAndVerticalOffset: 5%;

.follow-pick-marker {
  z-index: 2;
  width: 1px;
  min-width: 1px;
  opacity: var(--sd-opacity);

  & &__vertical {
    position: absolute;
    height: 90%;
    top: $labelAndVerticalOffset;
  }

  &__label {
    color: var(--follow-pick-marker-color, white);
    position: absolute;
    display: flex;
    align-items: center;
    font-size: 0.95rem;
    font-weight: 400;
    border-radius: 3px;
    line-height: 1;
    padding: 2px;
    z-index: 1;
    --label-background-color: rgba(25, 32, 37, 0.75);
    text-shadow: var(--text-shadow-color) 0px 0px 2px, var(--text-shadow-color) 1px 1px 3px,
      var(--text-shadow-color) 1px 1px 10px;

    &--top {
      top: $labelAndVerticalOffset;
      bottom: initial;
    }
    &--bottom {
      top: initial;
      bottom: $labelAndVerticalOffset;
    }
  }
}

.follow-pick-marker__vertical::after {
  content: '';
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  min-width: 1px;
  height: 100%;
  position: absolute;
  background-color: var(--follow-pick-marker-color, white);
  z-index: 3;
}
