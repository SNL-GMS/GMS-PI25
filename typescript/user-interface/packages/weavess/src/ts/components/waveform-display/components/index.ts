export { TimeAxis, YAxis } from './axes';
export {
  FollowPickMarker,
  MoveableMarker,
  PickMarker,
  SelectionWindow,
  VerticalMarker
} from './markers';
export { Station } from './station';
export { TheoreticalPhaseWindow } from './theoretical-phase-window';
