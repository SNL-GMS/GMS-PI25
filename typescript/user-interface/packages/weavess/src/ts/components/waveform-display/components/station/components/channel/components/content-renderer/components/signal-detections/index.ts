export { SignalDetection } from './signal-detection';
export { SignalDetections } from './signal-detections';
