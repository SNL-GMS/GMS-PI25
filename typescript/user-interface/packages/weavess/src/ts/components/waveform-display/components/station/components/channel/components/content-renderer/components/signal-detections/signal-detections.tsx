import isEqual from 'lodash/isEqual';
import memoizeOne from 'memoize-one';
import React from 'react';

import { SignalDetection } from './signal-detection';
import type { SignalDetectionsProps } from './types';

export class SignalDetections extends React.PureComponent<SignalDetectionsProps> {
  /**
   * Creates SignalDetection components
   *
   * @param props the signal detection props
   * @returns an array of signal detection elements as JSX.Element
   */
  private static readonly createSignalDetectionElements = (
    props: SignalDetectionsProps
  ): JSX.Element[] => {
    if (!props.signalDetections) return [];
    return props.signalDetections.map(signalDetection => {
      return (
        <SignalDetection
          key={`${signalDetection.id}-top`}
          stationId={props.stationId}
          channelId={props.channelId}
          signalDetection={signalDetection}
          displayInterval={props.displayInterval}
          viewableInterval={props.viewableInterval}
          offsetSecs={props.offsetSecs}
          events={props.events}
          isOnSplitChannel={props.isOnSplitChannel}
          initialConfiguration={props.initialConfiguration}
          // eslint-disable-next-line @typescript-eslint/unbound-method
          getTimeSecsForClientX={props.getTimeSecsForClientX}
          // eslint-disable-next-line @typescript-eslint/unbound-method
          getClientXForTimeSecs={props.getClientXForTimeSecs}
          // eslint-disable-next-line @typescript-eslint/unbound-method
          toggleDragIndicator={props.toggleDragIndicator}
          // eslint-disable-next-line @typescript-eslint/unbound-method
          positionDragIndicator={props.positionDragIndicator}
        />
      );
    });
  };

  /**
   * A memoized function for creating the signal detection elements.
   * The memoization function caches the results using
   * the most recent argument and returns the results.
   *
   * @param props the signal detection props
   * @returns an array JSX elements
   */
  private readonly memoizedCreateSignalDetectionElements: (
    props: SignalDetectionsProps
  ) => JSX.Element[];

  /**
   * Constructor
   *
   * @param props Waveform props as SignalDetectionsProps
   */
  public constructor(props: SignalDetectionsProps) {
    super(props);
    this.memoizedCreateSignalDetectionElements = memoizeOne(
      SignalDetections.createSignalDetectionElements,
      /* tell memoize to use a deep comparison for complex objects */
      isEqual
    );
    this.state = {};
  }

  public render(): JSX.Element {
    return <>{this.memoizedCreateSignalDetectionElements(this.props)}</>;
  }
}
