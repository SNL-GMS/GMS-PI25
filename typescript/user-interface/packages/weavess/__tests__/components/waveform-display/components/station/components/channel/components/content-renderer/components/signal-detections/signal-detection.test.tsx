/* eslint-disable react/jsx-props-no-spreading */
/* eslint-disable @typescript-eslint/no-magic-numbers */
import type { WeavessTypes } from '@gms/weavess-core';
import { WeavessConstants } from '@gms/weavess-core';
import { act, render } from '@testing-library/react';
import * as Enzyme from 'enzyme';
import * as React from 'react';

import { SignalDetection } from '../../../../../../../../../../../src/ts/components/waveform-display/components/station/components/channel/components/content-renderer/components/signal-detections';
import type { SignalDetectionProps } from '../../../../../../../../../../../src/ts/components/waveform-display/components/station/components/channel/components/content-renderer/components/signal-detections/types';
import { initialConfiguration } from '../../../../../../../../../../__data__/test-util-data';

const displayInterval: WeavessTypes.TimeRange = {
  startTimeSecs: 500,
  endTimeSecs: 600
};
const signalDetection: WeavessTypes.PickMarker = {
  /** unique id of the pick marker */
  id: 'foo',
  timeSecs: displayInterval.startTimeSecs + 10,
  uncertaintySecs: 1.5,
  showUncertaintyBars: true,
  label: 'P',
  color: 'red',
  isConflicted: false,
  isSelected: true,
  isDisabled: false,
  isActionTarget: false,
  isDraggable: true,
  filter: 'none'
};

const weavessEvents: WeavessTypes.ChannelContentEvents = {
  ...WeavessConstants.DEFAULT_UNDEFINED_EVENTS?.stationEvents?.defaultChannelEvents?.events,
  onSignalDetectionDragEnd: jest.fn()
};

const props: SignalDetectionProps = {
  stationId: 'my-station',
  channelId: 'my-channel',
  signalDetection,
  displayInterval,
  viewableInterval: displayInterval,
  offsetSecs: 0,
  events: weavessEvents,
  initialConfiguration,
  getTimeSecsForClientX: jest.fn(() => 2),
  toggleDragIndicator: jest.fn(),
  positionDragIndicator: jest.fn(),
  getClientXForTimeSecs: jest.fn(),
  isOnSplitChannel: false
};

const wrapper = Enzyme.mount(<SignalDetection {...props} />);
const instance: SignalDetection = wrapper.find(SignalDetection).instance() as SignalDetection;

describe('Weavess Signal Detection', () => {
  it('Weavess Pick Marker to be defined', () => {
    expect(SignalDetection).toBeDefined();
    expect(instance).toBeDefined();
  });

  it('renders', () => {
    const { container } = render(<SignalDetection {...props} />);
    expect(container).toMatchSnapshot();
  });
  it('set editing time uncertainty', async () => {
    expect((instance as any).state.isEditingTimeUncertainty).toBeFalsy();
    await act(() => {
      expect(() => (instance as any).setIsEditingTimeUncertainty(true)).not.toThrow();
    });
    expect((instance as any).state.isEditingTimeUncertainty).toBeTruthy();
  });

  it('set signal detection time', () => {
    expect(() =>
      (instance as any).setSignalDetectionTime(displayInterval.startTimeSecs + 50)
    ).not.toThrow();
  });

  it('set uncertainty in state', async () => {
    await act(() => {
      expect(() =>
        (instance as any).setUncertaintyInState(displayInterval.startTimeSecs + 25, true)
      ).not.toThrow();
    });
  });
});
