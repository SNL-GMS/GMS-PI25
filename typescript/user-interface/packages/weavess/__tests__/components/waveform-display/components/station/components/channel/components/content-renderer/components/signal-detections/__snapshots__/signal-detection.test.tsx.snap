// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weavess Signal Detection renders 1`] = `
<div>
  <div
    class="signal-detection__pick"
  >
    <div
      class="pick-marker pick-marker--selected pick-marker--selectable"
      style="--pick-marker-color: red;"
    >
      <div
        class="pick-marker__vertical"
        data-cy-predicted="pick-marker-pick-predicted-false"
        style="left: 10%; filter: none;"
      />
      <div
        class="pick-marker__label pick-marker__label--top"
        data-cy="pick-marker-foo"
        data-cy-color="pick-marker-red"
        data-cy-is-predicted-phase="false"
        data-cy-phase="P"
        data-cy-style-left="10"
        style="left: calc(4px + 10%); filter: none;"
      >
        P
      </div>
      <div
        class="uncertainty-marker"
      >
        <div
          class="uncertainty-marker__vertical uncertainty-marker__vertical--left"
          style="border-left: 1.5px solid red; left: 8.5%;"
        />
        <div
          class="uncertainty-marker__invisible-drag-target"
          style="left: 8.49%; right: 91.49%;"
        />
        <div
          class="uncertainty-marker__horizontal"
          style="border-top: 1.5px solid red; left: 8.5%; width: 1.5%;"
        />
      </div>
      <div
        class="uncertainty-marker"
      >
        <div
          class="uncertainty-marker__vertical uncertainty-marker__vertical--right"
          style="border-left: 1.5px solid red; left: 11.5%;"
        />
        <div
          class="uncertainty-marker__invisible-drag-target"
          style="left: 11.49%; right: 88.49%;"
        />
        <div
          class="uncertainty-marker__horizontal"
          style="border-top: 1.5px solid red; left: 10%; width: 1.5%;"
        />
      </div>
    </div>
  </div>
</div>
`;
